File: CoreLocation框架/OC_Code/05-指南针/05-指南针/ViewController.m
//
//  ViewController.m
//  05-指南针
//
//  Created by 1 on 15/12/21.
//  Copyright © 2015年 小码哥. All rights reserved.
//

#import "ViewController.h"
#import <CoreLocation/CoreLocation.h>
@interface ViewController ()<CLLocationManagerDelegate>

@property (weak, nonatomic) IBOutlet UIImageView *compassView;

/** 位置管理者*/
@property(nonatomic ,strong) CLLocationManager *locationM;


@end

@implementation ViewController


/**
 *  位置管理懒加载方法
 */
- (CLLocationManager *)locationM
{
    if (!_locationM) {
        _locationM = [[CLLocationManager alloc] init];
        _locationM.delegate = self;
    }
    return _locationM;
}


- (void)viewDidLoad {
    [super viewDidLoad];
    // 0. 判断当前磁力计传感器是否可用
    if (![CLLocationManager headingAvailable]) {
        return;
    }
    // 1. 获取当前设备朝向
    [self.locationM startUpdatingHeading];
    
    
    
}




#pragma mark - CLLocationManagerDelegate
/**
 *  当获取到设备朝向时
 *
 *  @param manager    位置管理者
 *  @param newHeading 头部信息
 */
-(void)locationManager:(CLLocationManager *)manager didUpdateHeading:(CLHeading *)newHeading
{
    if (newHeading.headingAccuracy < 0)
    {
        return;
    }
    
    // 1. 获取设备距离磁北方向的角度(0.0---359.9)
    CLLocationDirection angle = newHeading.magneticHeading;
    
    // 1.1 角度---> 弧度
    CGFloat hudu = angle / 180.0 * M_PI;
    
    // 2. 旋转图片(弧度)
    [UIView animateWithDuration:0.5 animations:^{
        self.compassView.transform = CGAffineTransformMakeRotation(-hudu);
    }];
    
    
}






@end

File: CoreLocation框架/Swift_Code/05-了解 - 指南针效果实现/05-了解 - 指南针效果实现/ViewController.swift
//
//  ViewController.swift
//  05-了解 - 指南针效果实现
//
//  Created by 王顺子 on 15/12/5.
//  Copyright © 2015年 小码哥. All rights reserved.
//

import UIKit
import CoreLocation

class ViewController: UIViewController, CLLocationManagerDelegate {

    @IBOutlet weak var compassView: UIImageView!


    // MARK: 懒加载
    lazy var locationM : CLLocationManager = {

        // 创建位置管理者对象
        let tempLocation : CLLocationManager = CLLocationManager()
        // 设置代理
        tempLocation.delegate = self
        return tempLocation

    }()

    override func viewDidLoad() {
        super.viewDidLoad()

        // 判断是否可以获取当前设备朝向
        if CLLocationManager.headingAvailable()
        {
            // 开始获取用户设备的朝向
            locationM.startUpdatingHeading()
        }

    }


    // MARK: CLLocationManagerDelegate
    func locationManager(manager: CLLocationManager, didUpdateHeading newHeading: CLHeading) {

        // 获取当前的手机朝向之前, 先判断数据是否有效
        if  newHeading.headingAccuracy < 0
        {
            return
        }

        // 1. 获取设备朝向(距离磁北的角度, 或者距离真北的角度)
        let angle = newHeading.magneticHeading

        // 1.1 转换成为弧度
        let radian = CGFloat(angle / 180.0 * M_PI)

        // 2. 反向旋转图片
        UIView.animateWithDuration(0.5) { () -> Void in
            self.compassView.transform = CGAffineTransformMakeRotation(-radian)
        };



    }



}

